File: server/parser.js
export class ParseError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ParseError';
  }
}

export function parse(buf) {
  if (buf.length < 323) {
    throw new ParseError(`packet too short: ${buf.length} bytes (need ≥323)`);
  }

  // Sled fields (bytes 0–231)
  const isRaceOn = buf.readInt32LE(0) !== 0;
  const timestampMs = buf.readUInt32LE(4);
  const engineMaxRpm = buf.readFloatLE(8);
  const engineIdleRpm = buf.readFloatLE(12);
  const currentEngineRpm = buf.readFloatLE(16);
  const accelX = buf.readFloatLE(20);
  const accelY = buf.readFloatLE(24);
  const accelZ = buf.readFloatLE(28);
  const velX = buf.readFloatLE(32);
  const velY = buf.readFloatLE(36);
  const velZ = buf.readFloatLE(40);
  
  // skip 3 floats AngularVelocity X/Y/Z (44, 48, 52)
  const yaw = buf.readFloatLE(56);
  const pitch = buf.readFloatLE(60);
  const roll = buf.readFloatLE(64);
  const suspensionFl = buf.readFloatLE(68);
  const suspensionFr = buf.readFloatLE(72);
  const suspensionRl = buf.readFloatLE(76);
  const suspensionRr = buf.readFloatLE(80);
  const tireSlipRatioFl = buf.readFloatLE(84);
  const tireSlipRatioFr = buf.readFloatLE(88);
  const tireSlipRatioRl = buf.readFloatLE(92);
  const tireSlipRatioRr = buf.readFloatLE(96);

  // skip 4 floats WheelRotationSpeed (100, 104, 108, 112)
  // skip 4 floats WheelOnRumbleStrip (116, 120, 124, 128)
  // skip 4 floats WheelInPuddleDepth (132, 136, 140, 144)
  // skip 4 floats SurfaceRumble (148, 152, 156, 160)

  const tireSlipAngleFl = buf.readFloatLE(164);
  const tireSlipAngleFr = buf.readFloatLE(168);
  const tireSlipAngleRl = buf.readFloatLE(172);
  const tireSlipAngleRr = buf.readFloatLE(176);

  // skip 4 floats TireCombinedSlip (180, 184, 188, 192)
  // skip 4 floats SuspensionTravelMeters (196, 200, 204, 208)

  const carOrdinal = buf.readInt32LE(212);
  const carClass = buf.readInt32LE(216);
  const carPi = buf.readInt32LE(220);
  const drivetrainType = buf.readInt32LE(224);
  const numCylinders = buf.readInt32LE(228);

  // Dash-only fields
  // skip bytes 232–243 (12 bytes)
  const positionX = buf.readFloatLE(244);
  const positionY = buf.readFloatLE(248);
  const positionZ = buf.readFloatLE(252);
  const speedMs = buf.readFloatLE(256);
  const power = buf.readFloatLE(260);
  const torque = buf.readFloatLE(264);

  // Game sends tire temps in Fahrenheit; convert to Celsius for display
  const tireTempFl = (buf.readFloatLE(268) - 32.0) * 5.0 / 9.0;
  const tireTempFr = (buf.readFloatLE(272) - 32.0) * 5.0 / 9.0;
  const tireTempRl = (buf.readFloatLE(276) - 32.0) * 5.0 / 9.0;
  const tireTempRr = (buf.readFloatLE(280) - 32.0) * 5.0 / 9.0;

  const boost = buf.readFloatLE(284);
  const fuel = buf.readFloatLE(288);
  const distanceTraveled = buf.readFloatLE(292);
  const bestLap = buf.readFloatLE(296);
  const lastLap = buf.readFloatLE(300);
  const currentLap = buf.readFloatLE(304);
  const currentRaceTime = buf.readFloatLE(308);
  
  const lapNumber = buf.readUInt16LE(312);
  const racePosition = buf.readUInt8(314);
  const throttle = buf.readUInt8(315);
  const brake = buf.readUInt8(316);
  const clutch = buf.readUInt8(317);
  const handbrake = buf.readUInt8(318);
  const gear = buf.readUInt8(319);
  const steer = buf.readInt8(320);
  const drivingLine = buf.readInt8(321);
  const aiBrakeDiff = buf.readInt8(322);

  // Optional tire wear (bytes 323+)
  let tireWearFl = null, tireWearFr = null, tireWearRl = null, tireWearRr = null;
  if (buf.length >= 327) tireWearFl = buf.readFloatLE(323);
  if (buf.length >= 331) tireWearFr = buf.readFloatLE(327);
  if (buf.length >= 335) tireWearRl = buf.readFloatLE(331);
  if (buf.length >= 339) tireWearRr = buf.readFloatLE(335);

  return {
    isRaceOn,
    timestampMs,
    engineMaxRpm,
    engineIdleRpm,
    currentEngineRpm,
    accelX,
    accelY,
    accelZ,
    velX,
    velY,
    velZ,
    positionX,
    positionY,
    positionZ,
    tireSlipRatioFl,
    tireSlipRatioFr,
    tireSlipRatioRl,
    tireSlipRatioRr,
    tireSlipAngleFl,
    tireSlipAngleFr,
    tireSlipAngleRl,
    tireSlipAngleRr,
    carOrdinal,
    carClass,
    carPi,
    drivetrainType,
    speedMs,
    power,
    torque,
    tireTempFl,
    tireTempFr,
    tireTempRl,
    tireTempRr,
    boost,
    fuel,
    distanceTraveled,
    bestLap,
    lastLap,
    currentLap,
    currentRaceTime,
    lapNumber,
    racePosition,
    throttle,
    brake,
    clutch,
    handbrake,
    gear,
    steer,
    yaw,
    pitch,
    roll,
    suspensionFl,
    suspensionFr,
    suspensionRl,
    suspensionRr,
    tireWearFl,
    tireWearFr,
    tireWearRl,
    tireWearRr,
  };
}

File: go-server/parser/parser.go
package parser

import (
	"encoding/binary"
	"errors"
	"math"
)

// float32FromBytes mengonversi 4 byte Little Endian menjadi float32
func float32FromBytes(b []byte) float32 {
	bits := binary.LittleEndian.Uint32(b)
	return math.Float32frombits(bits)
}

// int32FromBytes mengonversi 4 byte Little Endian menjadi int32
func int32FromBytes(b []byte) int32 {
	return int32(binary.LittleEndian.Uint32(b))
}

// Parse mendekode paket byte mentah (min 323 bytes) menjadi TelemetryPacket.
func Parse(buf []byte) (*TelemetryPacket, error) {
	if len(buf) < 323 {
		return nil, errors.New("paket terlalu pendek, pastikan format Forza Data Out disetel ke 'Dash'")
	}

	p := &TelemetryPacket{}

	p.IsRaceOn = binary.LittleEndian.Uint32(buf[0:4]) != 0
	p.TimestampMs = binary.LittleEndian.Uint32(buf[4:8])
	p.EngineMaxRpm = float32FromBytes(buf[8:12])
	p.EngineIdleRpm = float32FromBytes(buf[12:16])
	p.CurrentEngineRpm = float32FromBytes(buf[16:20])
	
	p.AccelX = float32FromBytes(buf[20:24])
	p.AccelY = float32FromBytes(buf[24:28])
	p.AccelZ = float32FromBytes(buf[28:32])
	p.VelX = float32FromBytes(buf[32:36])
	p.VelY = float32FromBytes(buf[36:40])
	p.VelZ = float32FromBytes(buf[40:44])

	// Skip 12 bytes AngularVelocity
	p.Yaw = float32FromBytes(buf[56:60])
	p.Pitch = float32FromBytes(buf[60:64])
	p.Roll = float32FromBytes(buf[64:68])

	p.SuspensionFl = float32FromBytes(buf[68:72])
	p.SuspensionFr = float32FromBytes(buf[72:76])
	p.SuspensionRl = float32FromBytes(buf[76:80])
	p.SuspensionRr = float32FromBytes(buf[80:84])

	p.TireSlipRatioFl = float32FromBytes(buf[84:88])
	p.TireSlipRatioFr = float32FromBytes(buf[88:92])
	p.TireSlipRatioRl = float32FromBytes(buf[92:96])
	p.TireSlipRatioRr = float32FromBytes(buf[96:100])

	// Skip Wheel Rotation, Rumble, Puddle, Surface (Total 64 bytes)
	p.TireSlipAngleFl = float32FromBytes(buf[164:168])
	p.TireSlipAngleFr = float32FromBytes(buf[168:172])
	p.TireSlipAngleRl = float32FromBytes(buf[172:176])
	p.TireSlipAngleRr = float32FromBytes(buf[176:180])

	// Skip CombinedSlip & SuspensionTravelMeters (Total 32 bytes)
	p.CarOrdinal = int32FromBytes(buf[212:216])
	p.CarClass = int32FromBytes(buf[216:220])
	p.CarPi = int32FromBytes(buf[220:224])
	p.DrivetrainType = int32FromBytes(buf[224:228])
	p.NumCylinders = int32FromBytes(buf[228:232])

	// Skip 12 bytes Unknown
	p.PositionX = float32FromBytes(buf[244:248])
	p.PositionY = float32FromBytes(buf[248:252])
	p.PositionZ = float32FromBytes(buf[252:256])
	p.SpeedMs = float32FromBytes(buf[256:260])
	p.Power = float32FromBytes(buf[260:264])
	p.Torque = float32FromBytes(buf[264:268])

	// Konversi suhu ban dari Fahrenheit ke Celsius langsung di Parser
	p.TireTempFl = (float32FromBytes(buf[268:272]) - 32.0) * 5.0 / 9.0
	p.TireTempFr = (float32FromBytes(buf[272:276]) - 32.0) * 5.0 / 9.0
	p.TireTempRl = (float32FromBytes(buf[276:280]) - 32.0) * 5.0 / 9.0
	p.TireTempRr = (float32FromBytes(buf[280:284]) - 32.0) * 5.0 / 9.0

	p.Boost = float32FromBytes(buf[284:288])
	p.Fuel = float32FromBytes(buf[288:292])
	p.DistanceTraveled = float32FromBytes(buf[292:296])
	p.BestLap = float32FromBytes(buf[296:300])
	p.LastLap = float32FromBytes(buf[300:304])
	p.CurrentLap = float32FromBytes(buf[304:308])
	p.CurrentRaceTime = float32FromBytes(buf[308:312])

	p.LapNumber = binary.LittleEndian.Uint16(buf[312:314])
	p.RacePosition = buf[314]
	p.Throttle = buf[315]
	p.Brake = buf[316]
	p.Clutch = buf[317]
	p.Handbrake = buf[318]
	p.Gear = buf[319]
	p.Steer = int8(buf[320])
	p.DrivingLine = int8(buf[321])
	p.AiBrakeDiff = int8(buf[322])

	return p, nil
}
